rfcs/006-mango-fdb.md

```diff
@@ -0,0 +1,154 @@
+
+### Indexes
+
+Each index defined in the Index Definition would have an index key space where the database’s documents are stored and sorted via the keys defined in the index’s definition. The data model for each defined index would be:
+
+`(?DATABASE, ?INDEXES, ?INDEX_NAME,  <indexed_field>, …<other indexed fields>, _id) = null`
+
+The `_id` is kept to avoid duplicate keys and to be used to retrieve the full document for a Mango query.
+For now, the value will be null, later we can look at storing covering indexes. aggregate values or materialised views.
+
+### Key sorting
+
+In CouchDB 2.x ICU collation is used to sort string key’s when added to the index’s b-tree. The current way of using ICU string collation won’t work with FoundationDB. To resolve this strings will be converted to an ICU sort string before being stored in FDB.  This is an extra performance overhead but will only be done when one when writing a key into the index.
+
+CouchDB has a defined [index collation specification](http://docs.couchdb.org/en/stable/ddocs/views/collation.html#collation-specification) that the new Mango design must adhere to. Each key added to a Mango index will be converted into a composite key or tuple with the first value in the tuple representing the type that the key so that it would be sorted correctly. Below is an example of the type keys to be used:
+
+\x00 NULL
+\x26 False
+\x27 True
+\x30 Numbers
+\x40 Text converted into a sort string
+\x50 Array
+\x60 Objects
+
+An example for a number key would be (\x30, 1). Note, Null and Boolean values won’t need to be composite keys as the type key is the value.
+
+### Index Limits
+
+This design has certain defined limits for it to work correctly:
+
+* The index definition (name, fields and partial_filter_selector) cannot exceed 100 KB FDB value limit
+* The sorted keys for an index cannot exceed the 10 KB key limit
+* To be able to update the index in the transaction that a document is updated in, there will have to be a limit on number of Mango indexes for a database so that the transaction stays within the 10MB transaction limit. This limit is still TBD based on testing.
+
+## Index building and management
+
+When an index is created on an existing database, the index will need to be built for all existing documents in the database. The process for building a new index would be:
+
+1. When a user defines a new index on an existing database, save the index definition along with the `sequence`  the index was added at and set the `build_status` to `building`  so it won’t be used to service queries. 
+2. Any write requests (document updates) after that must read the new index definition and update the index. When updating the new index, the index writers should assume that previous versions of the document have already been indexed.
```

Comment:
Can you elaborate to this? Say you have a transaction that does a write/update. In that same transaction, I'm assuming you want the index to incorporate this new write/update. What do you mean by "read the new index definition"?